## Laboratorio 03 - SQL parte II

#### Bienvenido

#### Davis Alvarez - 15842

Bienvenido a su tercer laboratorio de Bases de datos.

El objetivo de la actividad es que el estudiante profundice y practique sobre la aplicación de multi-table queries utilizando `INNER JOIN`s y `OUTER JOIN`s, además de utilizar funciones de agregación. El estudiante debe conocer la sintaxis de `SELECT` introduciendo los conceptos de agregación (`GROUP BY`) y filtros sobre datos agregados (`HAVING`). Adicionalmente se pretende introducir el uso de queries anidados ( _subqueries_ ). 

Iniciaremos cargando la extensión de python que nos permitirá interactuar con bases de datos SQLite:

In [1]:
%load_ext sql

### Consultas SQL (100 puntos)

Para este laboratorio utilizaremos una siguiente versión de la base de datos de vuelos, diseñada específicamente para este laboratorio: **flights-lab3.db**.

Puede descargar esta base de datos a partir de cualquier de las siguientes ubicaciones:

* Archivo comprimido en extensión .rar: https://drive.google.com/file/d/16b-Ppb5QlbA3K62qfRr-neGA-Ov5zzOG/view?usp=sharing
* Archivo comprimido en extensión .zip: https://drive.google.com/file/d/1vuhV7BlItkVOpaK_SiYI0Tmnu4tVwnCK/view?usp=sharing
* Archivo comprimido en extensión .tar.gz: https://drive.google.com/file/d/1mD0b2vsPE4dR1THM8OMOZReFPtsltKTp/view?usp=sharing

Recuerde que puede encontrar más información acerca de los campos en: https://www.transtats.bts.gov/DL_SelectFields.asp?Table_ID=236

In [2]:
sql sqlite:///flights-lab3.db

'Connected: @flights-lab3.db'

In [4]:
%sql PRAGMA table_info(ontime);

 * sqlite:///flights-lab3.db
Done.


[(0, 'Year', 'int', 0, None, 0),
 (1, 'Month', 'int', 0, None, 0),
 (2, 'DayofMonth', 'int', 0, None, 0),
 (3, 'DayOfWeek', 'int', 0, None, 0),
 (4, 'DepTime', 'int', 0, None, 0),
 (5, 'CRSDepTime', 'int', 0, None, 0),
 (6, 'ArrTime', 'int', 0, None, 0),
 (7, 'CRSArrTime', 'int', 0, None, 0),
 (8, 'UniqueCarrier', 'varchar(5)', 0, None, 0),
 (9, 'FlightNum', 'int', 0, None, 0),
 (10, 'TailNum', 'varchar(8)', 0, None, 0),
 (11, 'ActualElapsedTime', 'int', 0, None, 0),
 (12, 'CRSElapsedTime', 'int', 0, None, 0),
 (13, 'AirTime', 'int', 0, None, 0),
 (14, 'ArrDelay', 'int', 0, None, 0),
 (15, 'DepDelay', 'int', 0, None, 0),
 (16, 'Origin', 'varchar(3)', 0, None, 0),
 (17, 'Dest', 'varchar(3)', 0, None, 0),
 (18, 'Distance', 'int', 0, None, 0),
 (19, 'TaxiIn', 'int', 0, None, 0),
 (20, 'TaxiOut', 'int', 0, None, 0),
 (21, 'Cancelled', 'int', 0, None, 0),
 (22, 'CancellationCode', 'varchar(1)', 0, None, 0),
 (23, 'Diverted', 'varchar(1)', 0, None, 0),
 (24, 'CarrierDelay', 'int', 0, None, 0),
 (25, 'WeatherDelay', 'int', 0, None, 0),
 (26, 'NASDelay', 'int', 0, None, 0),
 (27, 'SecurityDelay', 'int', 0, None, 0),
 (28, 'LateAircraftDelay', 'int', 0, None, 0)]

In [5]:
%sql PRAGMA table_info(airports);

 * sqlite:///flights-lab3.db
Done.


[(0, 'iata', 'varchar(4)', 0, None, 0),
 (1, 'airport', 'varchar(100)', 0, None, 0),
 (2, 'city', 'varchar(100)', 0, None, 0),
 (3, 'state', 'varchar(4)', 0, None, 0),
 (4, 'country', 'varchar(4)', 0, None, 0),
 (5, 'lat', 'double', 0, None, 0),
 (6, 'long', 'double', 0, None, 0)]

Con esta base de datos cargada, proceda a preparar un query que responda a cada una de las siguientes preguntas:

#### Pregunta 1: ¿Cuáles son los vuelos más largo y más corto en distancia registrados en la base de datos? (10 puntos)

Responda a esta pregunta en un solo query:

In [9]:
%%sql SELECT MAX(distance)AS Larga, MIN(DISTANCE) AS Corta
        FROM ontime

 * sqlite:///flights-lab3.db
Done.


[(4962, 11)]

_Respuesta:_ La distancia más larga recorrida es de 4,962 millas, mientras que la más corta es de 11 millas.

#### Pregunta 2: ¿Qué aeropuertos no han tenido vuelos de salida? (10 puntos)

Para esta pregunta debe hacer uso de su conocimiento sobre `OUTER JOIN`s y valores `NULL`.

Su respuesta debe mostrar el código IATA, nombre y estado del aeropuerto.

In [41]:
%%sql 
SELECT A.iata, A.airport, A.state
FROM airports A
LEFT JOIN ontime O ON O.origin=A.iata 
WHERE O.origin IS NULL



 * sqlite:///flights-lab3.db
Done.


[('00M', 'Thigpen ', 'MS'),
 ('00R', 'Livingston Municipal', 'TX'),
 ('00V', 'Meadow Lake', 'CO'),
 ('01G', 'Perry-Warsaw', 'NY'),
 ('01J', 'Hilliard Airpark', 'FL'),
 ('01M', 'Tishomingo County', 'MS'),
 ('02A', 'Gragg-Wade ', 'AL'),
 ('02C', 'Capitol', 'WI'),
 ('02G', 'Columbiana County', 'OH'),
 ('03D', 'Memphis Memorial', 'MO'),
 ('04M', 'Calhoun County', 'MS'),
 ('04Y', 'Hawley Municipal', 'MN'),
 ('05C', 'Griffith-Merrillville ', 'IN'),
 ('05F', 'Gatesville - City/County', 'TX'),
 ('05U', 'Eureka', 'NV'),
 ('06A', 'Moton  Municipal', 'AL'),
 ('06C', 'Schaumburg', 'IL'),
 ('06D', 'Rolla Municipal', 'ND'),
 ('06M', 'Eupora Municipal', 'MS'),
 ('06N', 'Randall ', 'NY'),
 ('06U', 'Jackpot/Hayden ', 'NV'),
 ('07C', 'Dekalb County', 'IN'),
 ('07F', 'Gladewater Municipal', 'TX'),
 ('07G', 'Fitch H Beach', 'MI'),
 ('07K', 'Central City Municipal', 'NE'),
 ('08A', 'Wetumpka Municipal', 'AL'),
 ('08D', 'Stanley Municipal', 'ND'),
 ('08K', 'Harvard State', 'NE'),
 ('08M', 'Carthage-Leake County', 'MS'),
 ('09A', 'Butler-Choctaw County', 'AL'),
 ('09J', 'Jekyll Island', 'GA'),
 ('09K', 'Sargent Municipal', 'NE'),
 ('09M', 'Charleston Municipal', 'MS'),
 ('09W', 'South Capitol Street', 'DC'),
 ('0A3', 'Smithville Municipal', 'TN'),
 ('0A8', 'Bibb County', 'AL'),
 ('0A9', 'Elizabethton Municipal', 'TN'),
 ('0AK', 'Pilot Station', 'AK'),
 ('0B1', 'Col. Dyke ', 'ME'),
 ('0B4', 'Hartington Municipal', 'NE'),
 ('0B5', 'Turners Falls', 'MA'),
 ('0B7', 'Warren-Sugar Bush', 'VT'),
 ('0B8', 'Elizabeth ', 'NY'),
 ('0C0', 'Dacy', 'IL'),
 ('0C4', 'Pender Municipal', 'NE'),
 ('0D1', 'South Haven Municipal', 'MI'),
 ('0D8', 'Gettysburg Municipal', 'SD'),
 ('0E0', 'Moriarty', 'NM'),
 ('0E8', 'Crownpoint', 'NM'),
 ('0F2', 'Bowie Municipal', 'TX'),
 ('0F4', 'Loup City Municipal', 'NE'),
 ('0F7', 'Fountainhead Lodge Airpark', 'OK'),
 ('0F8', 'William R Pogue Municipal', 'OK'),
 ('0F9', 'Tishomingo Airpark', 'OK'),
 ('0G0', 'North Buffalo Suburban', 'NY'),
 ('0G3', 'Tecumseh Municipal', 'NE'),
 ('0G6', 'Williams County', 'OH'),
 ('0G7', 'Finger Lakes Regional', 'NY'),
 ('0H1', 'Trego Wakeeney ', 'KS'),
 ('0I8', 'Cynthiana-Harrison County', 'KY'),
 ('0J0', 'Abbeville Municipal', 'AL'),
 ('0J4', 'Florala Municipal', 'AL'),
 ('0J6', 'Headland Municipal', 'AL'),
 ('0K7', 'Humboldt Municipal', 'IA'),
 ('0L5', 'Goldfield', 'NV'),
 ('0L7', 'Jean', 'NV'),
 ('0L9', 'Echo Bay', 'NV'),
 ('0M0', 'Dumas Municipal', 'AR'),
 ('0M1', 'Scott ', 'TN'),
 ('0M4', 'Benton County', 'TN'),
 ('0M5', 'Humphreys County', 'TN'),
 ('0M6', 'Panola County', 'MS'),
 ('0M8', 'Byerley', 'LA'),
 ('0O3', 'Calaveras Co-Maury Rasmussen ', 'CA'),
 ('0O4', 'Corning Municipal', 'CA'),
 ('0O5', 'University', 'CA'),
 ('0Q5', 'Shelter Cove', 'CA'),
 ('0Q6', 'Shingletown', 'CA'),
 ('0R0', 'Columbia-Marion County', 'MS'),
 ('0R1', 'Atmore Municipal', 'AL'),
 ('0R3', 'Abbeville Chris Crusta Memorial', 'LA'),
 ('0R4', 'Concordia Parish', 'LA'),
 ('0R5', 'David G Joyce', 'LA'),
 ('0R7', 'Red River', 'LA'),
 ('0S7', 'Dorothy Scott', 'WA'),
 ('0S9', 'Jefferson County International', 'WA'),
 ('0V2', 'Harriet Alexander ', 'CO'),
 ('0V3', 'Pioneer Village ', 'NE'),
 ('0V4', 'Brookneal/Campbell County', 'VA'),
 ('0V6', 'Mission Sioux', 'SD'),
 ('0V7', 'Kayenta', 'AZ'),
 ('10C', 'Galt', 'IL'),
 ('10D', 'Winsted Municipal', 'MN'),
 ('10G', 'Holmes County', 'OH'),
 ('10N', 'Wallkill', 'NY'),
 ('10U', 'Owyhee', 'NV'),
 ('11A', 'Clayton Municipal', 'AL'),
 ('11D', 'Clarion Cty', 'PA'),
 ('11IS', 'Schaumburg Heliport', 'IL'),
 ('11J', 'Early County', 'GA'),
 ('11R', 'Brenham Municipal', 'TX'),
 ('12C', 'Rochelle Municipal', 'IL'),
 ('12D', 'Tower Municipal', 'MN'),
 ('12J', 'Brewton Municipal', 'AL'),
 ('12K', 'Superior Municipal', 'NE'),
 ('12Y', 'Le Sueur Municipal', 'MN'),
 ('13C', 'Lakeview', 'MI'),
 ('13K', 'Eureka Municipal', 'KS'),
 ('13N', 'Trinca', 'NJ'),
 ('14J', 'Carl Folsom', 'AL'),
 ('14M', 'Hollandale Municipal', 'MS'),
 ('14Y', 'Todd Field ', 'MN'),
 ('15F', 'Haskell Municipal', 'TX'),
 ('15J', '

Si hacemos la cuenta de las filas nos da da 3007 aeropuertos:

In [42]:
%%sql 
SELECT COUNT(*)
FROM airports A
LEFT JOIN ontime O ON O.origin=A.iata 
WHERE O.origin IS NULL

 * sqlite:///flights-lab3.db
Done.


[(3007,)]

_Hint_ : En la base de datos existen 3007 aeropuertos para los que no hay vuelos de salida

#### Pregunta 3: ¿Cuál es la temporada alta?

Escriba un query que muestre cuáles son los dos meses con más vuelos en la base de datos

In [47]:
%%sql SELECT month, count(origin) as vuelos
        FROM ontime
        GROUP BY month ORDER BY vuelos DESC
        LIMIT 2

 * sqlite:///flights-lab3.db
Done.


[(7, 627931), (3, 616090)]

#### Pregunta 4: ¿Cuáles son los peores días para viajar? Parte 1 (10 puntos)

Prepare un query que muestre el promedio de retraso en el tiempo de llegada (arrival) registrado para cada uno de los días de la semana utilizando funciones de agregación.

El esquema de la relación obtenida debe ser `(dia: int, retraso_promedio: float)` y debe mostrar los días ordenados por el  peor día para viajar al mejor.

In [48]:
%%sql SELECT DayOfWeek, AVG(ArrDelay) as promedio
        FROM ontime
        GROUP BY DayOfWeek ORDER BY promedio DESC

 * sqlite:///flights-lab3.db
Done.


[(5, 10.675214410055972),
 (7, 9.294087238339747),
 (4, 8.246549198623555),
 (1, 8.030254747872275),
 (2, 7.29232332960935),
 (3, 6.379804071503802),
 (6, 5.682499626837823)]

_Respuesta_ :

```
5 	10.675214410055972
7 	9.294087238339747
4 	8.246549198623555
1 	8.030254747872275
2 	7.29232332960935
3 	6.379804071503802
6 	5.682499626837823
```

#### Pregunta 5: ¿Cuáles son los peores días para viajar? Parte 2 (10 puntos)

Dado que el resultado anterior no necesariamente es significativo para un usuario final, es necesario unir la información sobre el día del vuelo con la tabla `weekdays` contenida en esta nueva versión de la base de datos.

Puede investigar más sobre esta tabla utilizando un Jupyter notebook paralelo a manera de _scratchpad_.

A continuación trabaje el query necesario para responder a la pregunta sobre los peores días para viajar de tal forma que su respuesta incluya el nombre del día y el promedio de tiempo de retraso, `(dia_nombre: int, retraso_promedio: float)`:

In [51]:
%sql select * from weekdays

 * sqlite:///flights-lab3.db
Done.


[(1, 'Monday'),
 (2, 'Tuesday'),
 (3, 'Wednesday'),
 (4, 'Thursday'),
 (5, 'Friday'),
 (6, 'Saturday'),
 (7, 'Sunday')]

In [57]:
%%sql SELECT weekday_name AS Dia, AVG(ArrDelay) as promedio
        FROM ontime O INNER JOIN weekdays W ON W.DayofWeek= O.DayOfWeek
        GROUP BY O.DayOfWeek ORDER BY promedio DESC

 * sqlite:///flights-lab3.db
Done.


[('Friday', 10.675214410055972),
 ('Sunday', 9.294087238339747),
 ('Thursday', 8.246549198623555),
 ('Monday', 8.030254747872275),
 ('Tuesday', 7.29232332960935),
 ('Wednesday', 6.379804071503802),
 ('Saturday', 5.682499626837823)]

_Respuesta_ :

```
Friday 	10.675214410055972
Sunday 	9.294087238339747
Thursday 	8.246549198623555
Monday 	8.030254747872275
Tuesday 	7.29232332960935
Wednesday 	6.379804071503802
Saturday 	5.682499626837823
```

#### Pregunta 6: ¿Cuál es el vuelo más común? (15 puntos)

Considerando como vuelos repetidos aquellos que van de un mismo aeropuerto origen a un mismo aeropuerto destino, escriba un query que retorne los cinco vuelos más frecuentes registrados en la base de datos.

Por ejemplo, hay 11,224 vuelos registrados en la base de datos cuyo origen es `SAN` y cuyo destino es `LAX`.

In [11]:
%%sql SELECT origin, dest, COUNT(FlightNum) AS cuenta
        FROM ontime
        GROUP BY origin, dest
        ORDER BY cuenta DESC
        LIMIT 5

 * sqlite:///flights-lab3.db
Done.


[('SFO', 'LAX', 13788),
 ('LAX', 'SFO', 13390),
 ('OGG', 'HNL', 12383),
 ('LGA', 'BOS', 12035),
 ('BOS', 'LGA', 12029)]

#### Pregunta 7: ¿Cuál es el estado con más vuelos internos? (15 puntos)

Escriba un query que indique cuál es el estado con más vuelos internos.

Un vuelo interno es aquel cuyo aeropuerto origen y aeropuerto destino se encuentran en el mismo estado.

_Hint_ : Necesitará hacer dos JOINs a la tabla `airport` con condiciones diferentes. Inicie preparando un query que le muestre 10 ejemplos de vuelos internos, para verificar que la lógica de su query devuelve tuplas válidas.

In [29]:
%%sql SELECT  COUNT(O.FlightNum) AS cuenta, ORIGEN.state AS origen, DESTINO.state as destino
        FROM ontime O
            INNER JOIN airports ORIGEN ON O.origin=ORIGEN.iata
            INNER JOIN airports DESTINO ON O.dest=DESTINO.iata
        WHERE origen=destino
        GROUP BY origen, destino
        ORDER BY cuenta DESC
        LIMIT 1

 * sqlite:///flights-lab3.db
Done.


[(302086, 'CA', 'CA')]

_Respuesta:_ El estado con más vuelos internos es California (CA) con 302,086 vuelos internos

In [13]:
%sql PRAGMA table_info(airports);

 * sqlite:///flights-lab3.db
Done.


[(0, 'iata', 'varchar(4)', 0, None, 0),
 (1, 'airport', 'varchar(100)', 0, None, 0),
 (2, 'city', 'varchar(100)', 0, None, 0),
 (3, 'state', 'varchar(4)', 0, None, 0),
 (4, 'country', 'varchar(4)', 0, None, 0),
 (5, 'lat', 'double', 0, None, 0),
 (6, 'long', 'double', 0, None, 0)]

#### Pregunta 8: ¿Qué aerolíneas se retrasan más de 10 minutos regularmente? (10 puntos)

Escriba un query que retorne las aerolíneas cuyos vuelos se retrasan en promedio 10 minutos o más, ordenando de la que más se retrasa a la que menos.

El esquema de la relación obtenida debe ser de la forma `(airline_name: varchar, average_delay: float)`.

Hint: el uso de la cláusula `HAVING` puede resultar conveniente. Para la tabla de vuelos, utilice el atributo `UniqueCarrier` para asociar con el código de la aerolinea.

In [40]:
%%sql SELECT airlines.airlineName, AVG(ArrDelay) as retraso
        FROM ontime
            INNER JOIN airlines ON ontime.UniqueCarrier= Airlines.AirlineCode
        GROUP BY UniqueCarrier
            HAVING AVG(ArrDelay)>=10.0
        ORDER BY retraso DESC 

 * sqlite:///flights-lab3.db
Done.


[('American Airlines Inc.', 12.202853434950445),
 ('PSA Airlines Inc.', 11.404110178283158),
 ('Mesa Airlines Inc.', 11.322566979170753),
 ('United Air Lines Inc.', 11.001550560048052),
 ('JetBlue Airways', 10.859381613638567),
 ('Continental Air Lines Inc.', 10.809820575966226),
 ('ExpressJet Airlines Inc. (1)', 10.320298523403915),
 ('ExpressJet Airlines Inc.', 10.00033146217589)]

_Respuesta:_

```
American Airlines Inc. 	12.202853434950445
PSA Airlines Inc. 	11.404110178283158
Mesa Airlines Inc. 	11.322566979170753
United Air Lines Inc. 	11.001550560048052
JetBlue Airways 	10.859381613638567
Continental Air Lines Inc. 	10.809820575966226
ExpressJet Airlines Inc. (1) 	10.320298523403915
ExpressJet Airlines Inc. 	10.00033146217589
```

#### Pregunta 9: ¿Qué aerolíneas de las que vuelan desde SFO se retrasan más en llegar? (10 puntos)

Responderemos a esta pregunta en dos partes:

Para comenzar, escriba un query que retorne el listado de aerolineas que realizaron al menos un vuelo desde el aeropuerto `SFO`.

Su respuesta debe incluir únicamente el código de la aerolinea:

In [ ]:
%%sql SELECT 
        FROM 
        WHERE 

In [12]:
%sql PRAGMA table_info(ontime);

 * sqlite:///flights-lab3.db
Done.


[(0, 'Year', 'int', 0, None, 0),
 (1, 'Month', 'int', 0, None, 0),
 (2, 'DayofMonth', 'int', 0, None, 0),
 (3, 'DayOfWeek', 'int', 0, None, 0),
 (4, 'DepTime', 'int', 0, None, 0),
 (5, 'CRSDepTime', 'int', 0, None, 0),
 (6, 'ArrTime', 'int', 0, None, 0),
 (7, 'CRSArrTime', 'int', 0, None, 0),
 (8, 'UniqueCarrier', 'varchar(5)', 0, None, 0),
 (9, 'FlightNum', 'int', 0, None, 0),
 (10, 'TailNum', 'varchar(8)', 0, None, 0),
 (11, 'ActualElapsedTime', 'int', 0, None, 0),
 (12, 'CRSElapsedTime', 'int', 0, None, 0),
 (13, 'AirTime', 'int', 0, None, 0),
 (14, 'ArrDelay', 'int', 0, None, 0),
 (15, 'DepDelay', 'int', 0, None, 0),
 (16, 'Origin', 'varchar(3)', 0, None, 0),
 (17, 'Dest', 'varchar(3)', 0, None, 0),
 (18, 'Distance', 'int', 0, None, 0),
 (19, 'TaxiIn', 'int', 0, None, 0),
 (20, 'TaxiOut', 'int', 0, None, 0),
 (21, 'Cancelled', 'int', 0, None, 0),
 (22, 'CancellationCode', 'varchar(1)', 0, None, 0),
 (23, 'Diverted', 'varchar(1)', 0, None, 0),
 (24, 'CarrierDelay', 'int', 0, None, 0),
 (25, 'WeatherDelay', 'int', 0, None, 0),
 (26, 'NASDelay', 'int', 0, None, 0),
 (27, 'SecurityDelay', 'int', 0, None, 0),
 (28, 'LateAircraftDelay', 'int', 0, None, 0)]

In [35]:
%%sql SELECT *
        FROM airlines
        LIMIT 5

 * sqlite:///flights-lab3.db
Done.


[('02Q', 'Titan Airways'),
 ('04Q', 'Tradewind Aviation'),
 ('05Q', 'Comlux Aviation, AG'),
 ('06Q', 'Master Top Linhas Aereas Ltd.'),
 ('07Q', 'Flair Airlines Ltd.')]

A continuación investigue el uso del operador `IN` dentro de una clausula `WHERE` para identificar cuando un atributo se encuentra dentro de un multiset.

Luego encuentre el promedio de retraso de todas las aerolineas **cuyo nombre se encuentre dentro del resultado del query anterior**.

Su respuesta debe incluir el código de la aerolinea y el promedio en el tiempo de llegada.

_Hint_ : Utilice un _subquery_ como parte de la clausula `WHERE`.

In [ ]:
%%sql SELECT 
        FROM 
        WHERE 

_Respuesta_ :

```
AA 	12.202853434950445
UA 	11.001550560048052
B6 	10.859381613638567
CO 	10.809820575966226
XE 	10.320298523403915
MQ 	9.496970610952266
FL 	8.988157472371256
DL 	7.716164635751918
NW 	7.293465879672776
OO 	6.43893863978179
F9 	6.084135669681085
WN 	5.115703380225903
AS 	4.721360405553864
US 	2.800998260539828
HA 	1.2518519716624075
```